### PR TITLE
Max workers: AggregatedPool

### PR DESCRIPTION
**Pull request by kenotron**
We now move to using a worker pool aggregator that will group workers based on some passed-in function. In the CLI / Scheduler front, we decide to use the group function that determines groups based "task" name. 

The way to do this is to allow users to put in "maxWorkers" options for each task type. Then, if the concurrency is < the total sum of the maxWorkers, we then have to give the maxWorkers some adjustments based on the ratio of workers / sum of workers * concurrency. This results in scaling the # of workers down to fit the concurrency CLI arg (defaults to CPU cores - 1). If even that kind of scaling is more than the concurrency, we throw an error.

### CODE REVIEW

File: packages/cli/src/commands/run/filterPipelineDefinitions.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import type { PipelineDefinition } from "../../types/PipelineDefinition";
+
+export function filterPipelineDefinitions(targets: IterableIterator<Target>, pipeline: PipelineDefinition) {
+  const tasksSet = new Set<string>();
+
+  for (const target of targets) {
+    tasksSet.add(target.task);
+  }
+  const filteredPipeline: PipelineDefinition = {};
+  for (const [id, definition] of Object.entries(pipeline ?? {})) {
```
Comment (kerrynf) [inline packages/cli/src/commands/run/filterPipelineDefinitions.ts:11]: pipeline ?? {}

I'm missing something here - what happens when the pipeline is null?

Comment (kenotron) [reply packages/cli/src/commands/run/filterPipelineDefinitions.ts:11]: pipeline being null would just set some defaults for you. default mode for tasks is "task: [^task]" I think

File: packages/worker-threads-pool/src/AggregatedPool.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    for (const [group, groupMaxWorkers] of maxWorkersByGroup.entries()) {
+      const pool = new WorkerPool({ maxWorkers: groupMaxWorkers, workerOptions, script });
+      this.groupedPools.set(group, pool);
+      totalGroupedWorkers += groupMaxWorkers;
+    }
+
+    if (totalGroupedWorkers > maxWorkers) {
+      throw new Error(
+        `Total maxWorkers (${totalGroupedWorkers}) configured across all groups exceeds concurrency (${maxWorkers}). Hint: use percent of concurrency to configure the maxWorkers like this: { maxWorkers: "50%" }`
+      );
```
Comment (kerrynf) [inline packages/worker-threads-pool/src/AggregatedPool.ts:35]: Didn't you just outlaw strings in the config, or is that a different setting?

Comment (kerrynf) [reply packages/worker-threads-pool/src/AggregatedPool.ts:35]: (if you support % it must be a different setting)

**Review by kerrynf — APPROVED** [review]
It would be nice to have a unit test for AggregatePool? :)

File: packages/worker-threads-pool/src/AggregatedPool.ts (hunk 2 above)
Comment (kenotron) [reply packages/worker-threads-pool/src/AggregatedPool.ts:35]: yes, added some

**Pull request merged**